File: Java & Python/trees/flatten_tree_linked_list.java
/* preorder array of tree, node = queue.remove(), node = node.right */
public class Solution {
	public TreeNode flatten(TreeNode A) {
		Queue<TreeNode> queue = new LinkedList<TreeNode>();
		preorder(queue, A);

		if (queue.isEmpty())
			return A;

		TreeNode node = queue.remove();
		A = node;
		while (!queue.isEmpty()) {
			node.left = null;
			node.right = queue.remove();
			node = node.right;
		}
		return A;
	}


	// Let's try non recursive way
	public void preorder(Queue<TreeNode> queue, TreeNode node) {
		if (node == null)
			return;
		queue.add(node);
		preorder(queue, node.left);
		preorder(queue, node.right);
	}
}

/*
Given a binary tree, flatten it to a linked list in-place.

Example :
Given


         1
        / \
       2   5
      / \   \
     3   4   6
The flattened tree should look like:

   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6
Note that the left child of all nodes should be NULL.
*/

File: Java & Python/trees/flatten_tree_linked_list.py
# if root.left, go down to its right most child (Node X)
# attach everything from the right side to Node X
# shift root.left over to root.right

class Solution:
	# @param A : root node of tree
	# @return the root node in the tree
	def flatten(self, A):
		node = root = A
		while root:
			if root.left:
				prev = root.left
				while prev.right:
					prev = prev.right
				prev.right = root.right
				root.right = root.left
				root.left = None
			root = root.right
		return node

"""
Flatten Binary Tree to Linked ListBookmark
Given a binary tree, flatten it to a linked list in-place.

Example :
Given


         1
        / \
       2   5
      / \   \
     3   4   6
The flattened tree should look like:

   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6
Note that the left child of all nodes should be NULL.
"""